Src/Support/Google.Apis.Auth/OAuth2/TokenRefreshManager.cs

```diff
@@ -0,0 +1,170 @@
+limitations under the License.
+*/
+
+using Google.Apis.Auth.OAuth2.Responses;
+using Google.Apis.Logging;
+using Google.Apis.Util;
+using System;
+using System.Linq;
+using System.Runtime.ExceptionServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Google.Apis.Auth.OAuth2
+{
+    /// <summary>
+    /// Encapsulation of token refresh behaviour. This isn't entirely how we'd design the code now (in terms of the
+    /// callback in particular) but it fits in with the exposed API surface of ServiceCredential and UserCredential.
+    /// </summary>
+    internal sealed class TokenRefreshManager
+    {
+        // Immutable state
+        private readonly object _lock = new object();
+        private readonly IClock _clock;
+        private readonly ILogger _logger;
+        private readonly Func<CancellationToken, Task<bool>> _refreshAction;
+
+        // Mutable state, guarded with _lock.
+        private TokenResponse _token;
+        private Task<TokenResponse> _refreshTask;
+
+        /// <summary>
+        /// Creates a manager which executes the given refresh action when required.
+        /// </summary>
+        /// <param name="refreshAction">The refresh action which will populate the Token property when successful.</param>
+        /// <param name="clock">The clock to consult for timeouts.</param>
+        /// <param name="logger">The logger to use to record refreshes.</param>
+        internal TokenRefreshManager(Func<CancellationToken, Task<bool>> refreshAction, IClock clock, ILogger logger)
+        {
+            _refreshAction = refreshAction;
+            _clock = clock;
+            _logger = logger;
+        }
+
+        internal TokenResponse Token
+        {
+            get
+            {
+                lock (_lock)
+                {
+                    return _token;
+                }
+            }
+            // The token may be set due to operations other than GetAccessTokenForRequestAsync, but we don't need to
+            // null out _refreshTask if so; it will be 
+            set
+            {
+                lock (_lock)
+                {
+                    _token = value;
+                }
+            }
+        }
+
+        internal async Task<string> GetAccessTokenForRequestAsync(string authUri, CancellationToken cancellationToken)
+        {
+            TokenResponse currentToken;
+            Task<TokenResponse> currentRefreshTask;
+            lock (_lock)
+            {
+                currentToken = _token;
+                currentRefreshTask = _refreshTask;
+            }
+            // Note: IsExpired is really "should be refreshed".
+            if (currentToken == null || currentToken.IsExpired(_clock))
+            {
+                // Start a refresh if nothing else is already refreshing the token.
+                // It's possible to end up with multiple requests here (if another thread started a task while
```

Comment:
I'm fairly sure this can be avoided with a bit of code re-arrangement here.
As far as I can see, enough code needs to be inside the above `lock` statement, such that `_refreshTask` is set (if required) before the `lock` exist.
There doesn't seem to be any blocking code between the end of the lock and the `_refreshTask = currentRefreshTask` statement. The call to `RefreshTokenAsync()` could be wrapped in `Task.Run` if necessary.
I agree this isn't exactly critical, but it'd be nice to get it correct if possible.

Reply:
RefreshTokenAsync() calls "somewhat arbitrary" code (a delegate, but one within this library - at the moment); I'd prefer not to do that within a lock if we can help it. Will give it a try, remembering whether or not we need to kick off another task.

Reply:
Mmm... looked again, and if we *do* need to start a new task, we'd need to call the method in order to get the task to store within the lock. I'd rather not use Task.Run just to avoid blocking - taking out the lock twice feels more reasonable, and I don't think it's a problem to have more than one refresh request, so long as it's limited in the way we've got here. Happy to discuss more though.